walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,91 @@
 
 package com.google.sps;
 
+import java.util.Arrays;
+import java.util.ArrayList; 
 import java.util.Collection;
+import java.util.Collections;
+import java.util.List; 
+
 
 public final class FindMeetingQuery {
-  public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
-  }
+    public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
+        Collection<String> meeting_attendees = request.getAttendees();
+        if (meeting_attendees.size() <= 0) {
+            return Arrays.asList(TimeRange.WHOLE_DAY);
+        }
+        else if (request.getDuration() > TimeRange.WHOLE_DAY.duration()) {
+            return Arrays.asList();
+        }
+        else {
+            List<TimeRange> busy_times = getBusyTimes(events, meeting_attendees);
+
+            // sort busy_times by start time, at each time range split from before and after
+            Collections.sort(busy_times, TimeRange.ORDER_BY_START);
+            
+            List<TimeRange> all_available_times = getAllAvailableTimes(busy_times);
+            return getUsableTimes(all_available_times, request.getDuration());
+        }
+        // throw new UnsupportedOperationException("TODO: Implement this method.");
+    }
+
+    /** Returns list of TimeRanges when at least one meeting attendee is busy */
+    private List<TimeRange> getBusyTimes(Collection<Event> events, Collection<String> meeting_attendees) {
+        List<TimeRange> busy_times = new ArrayList<TimeRange>();
+        for (Event event : events) {
+            Collection<String> overlap_attendees = event.getAttendees();
+            overlap_attendees.retainAll(meeting_attendees);
+            if (overlap_attendees.size() > 0) {
+                busy_times.add(event.getWhen());
+            }
+        }
+        return busy_times;
+    }
+
+    /**
+     * Returns all available (non-busy) time slots in a day
+     * busy_times: sorted by start time
+     */
+    private List<TimeRange> getAllAvailableTimes(List<TimeRange> busy_times) {
+        List<TimeRange> available_times = new ArrayList<TimeRange>();
+        if (busy_times.size() <= 0) {
+            available_times.add(TimeRange.WHOLE_DAY);
+            return available_times;
+        }
+
+        int start = TimeRange.START_OF_DAY;
+        int end;
+        for (TimeRange busy_slot : busy_times) {
+            end = busy_slot.start();
+            // Case 1: end comes before start
+            if (end < start) {
```

Comment:
Seeing `end < start` is a bit cognitively jarring.
I'm wondering if better names could be found, but I'm at a bit of a loss.
Maybe `free_start` and `busy_start`?

I had to draw diagrams and work through the code to see what it was doing, so I'm trying to find a way to make it clearer what is being done here.